### PR TITLE
[ClangImporter] Track declaration context associated with type converter

### PR DESCRIPTION
**Pull request by xedin**
Objective-C allows to declare `typedef` inside interface
which is allowed to reference its generic parameters and
used by class methods, when importing types to Swift let's
track current declaration context and see if we've
encountered such situation, if so we are going to treat
such `typedef` as bound type of the generic paramter.

Resolves: rdar://problem/32298667

### CODE REVIEW

**Comment by xedin** [comment]
@jrose-apple Please take a look and let me know if it makes any sense to you...

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ImportType.cpp [hunk 1]
```diff
@@ -590,9 +584,21 @@ namespace {
...
       }
+
       if (index > genericSig->getGenericParams().size()) {
         return ImportResult();
       }
+
+      // Check if parameter's declaration context is the same as
+      // current declaration context, otherwise we can't use this
+      // type. This might happen when Objective-C `typedef` is
+      // declarated instead of interface and tries to use its
```
Comment (slavapestov) [inline lib/ClangImporter/ImportType.cpp]: "declared inside of an interface"?

File: test/ClangImporter/Inputs/typedef-with-generic-param.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#import <Foundation/Foundation.h>
+
+@interface Foo <__covariant T>
+typedef void (^CompletionBlock)(T _Nullable result, NSError *_Nullable error);
```
Comment (slavapestov) [inline test/ClangImporter/Inputs/typedef-with-generic-param.h]: Why are we unable to import this typedef though? It should be no different than the Swift code and the ClangImporter should be able to construct it without problems:

```
class Foo<T> {
   typealias CompletionBlock = () throws -> T
}
```

Comment (xedin) [inline test/ClangImporter/Inputs/typedef-with-generic-param.h]: The problem here is not `typedef` itself but how it's used in the class method down the road.

File: lib/ClangImporter/ImportType.cpp (hunk 1 above)
Comment (xedin) [inline lib/ClangImporter/ImportType.cpp]: Ah damn, I knew that I'm going to screw this up :(

**Comment by xedin** [comment]
@jrose-apple I've added `DC` tracking to `SwiftDeclConverter` as well because `VisitTypedefType` needs to build type decl, which requires correct context to be present.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ImportType.cpp [hunk 3]
```diff
@@ -590,12 +602,25 @@ namespace {
...
+      auto type = genericEnv->mapTypeIntoContext(paramDecl);
+
+      // Check if parameter's declaration context is the same as
+      // current declaration context, otherwise we can't use type
+      // type without re-substituting it. This might happen when
+      // Objective-C `typedef` is declarated inside of interface
+      // and tries to use its generic parameters.
+      if (auto *ownerDC = genericEnv->getOwningDeclContext()) {
+        if (DC && DC != ownerDC && !ownerDC->isChildContextOf(DC))
+          type = mapTypeIntoContext(ownerDC, DC, type);
```
Comment (DougGregor) [inline lib/ClangImporter/ImportType.cpp]: This seems more complicate than necessary... can't we just `DC->mapTypeIntoContext(type)`  once, and not bother doing the `DC != ownerDC` check?

**Pull request closed without merging**